src/org/jetbrains/java/decompiler/modules/decompiler/flow/FlattenStatementsHelper.java

```diff
@@ -80,722 +97,553 @@ private DirectNode createDirectNode(Statement stat, DirectNodeType type) {
-                } else {
-                  if (!catchall.containsStatementStrict(destination)) {
-                    stack.removeLast();
-                    created = false;
-                  } else {
-                    saveEdge(sourcenode, destination, edgetype, isFinallyExit ? finallyShortRangeSource : null, finallyLongRangeSource,
-                      finallyShortRangeEntry, finallyLongRangeEntry, isFinallyMonitorExceptionPath);
-                  }
-                }
-              } else { // finally protected try statement
-                if (!catchall.containsStatementStrict(destination)) {
-
-                  // FIXME: this is a hack, the edges need to be more properly defined from the finally handler to it's destination
-                  //  Otherwise problems can occur where variable usage scopes aren't correct!
-                  // Edge from finally handler head to destination
-                  listEdges.add(new Edge(sourcenode.id, destination.id, edgetype));
-
-                  saveEdge(sourcenode, catchall.getHandler(), StatEdge.TYPE_REGULAR, isFinallyExit ? finallyShortRangeSource : null,
-                    finallyLongRangeSource, finallyShortRangeEntry, finallyLongRangeEntry, isFinallyMonitorExceptionPath);
-
-                  stack.removeLast();
-                  stack.add(new StackEntry(catchall, true, edgetype, destination, catchall.getHandler(),
-                    finallyLongRangeEntry == null ? catchall.getHandler() : finallyLongRangeEntry,
-                    sourcenode, finallyLongRangeSource, false));
-
-                  statEntry.edgeIndex = edgeindex + 1;
-                  statEntry.succEdges = lstSuccEdges;
-                  lstStackStatements.addFirst(statEntry);
-                  lstStackStatements.addFirst(new StatementStackEntry(catchall.getHandler(), stack, null));
-
-                  continue mainloop;
-                } else {
-                  saveEdge(sourcenode, destination, edgetype, isFinallyExit ? finallyShortRangeSource : null, finallyLongRangeSource,
-                    finallyShortRangeEntry, finallyLongRangeEntry, isFinallyMonitorExceptionPath);
+        for (int i = 1; i < stat.getStats().size(); i++) {
+          this.flattenStatement(stat.getStats().get(i));
+        }
+
+        // add edges
+        this.addEdges(outNode, first.getAllDirectSuccessorEdges());
+
+        if (ifStat.iftype == IfStatement.IFTYPE_IF && stat.hasAnyDirectSuccessor()) {
+          // add implicit else edge
+          this.addEdge(outNode, stat.getFirstDirectSuccessor());
+        }
+
+        // Adds an edge from the last if statement to the current if statement,
+        // if the current if statement's head statement has no predecessor
+        // This was made to mask a failure in EliminateLoopsHelper and isn't used currently
+        // (over the current test set) but could theoretically still happen!
+        // TODO: what?
+        // TODO: use java code gen to generate a test for this?
```

Comment:
Just remove this code block, I don't think it's needed anymore

Reply:
ok